core/src/main/java/org/apache/iceberg/BaseTableScan.java

```diff
@@ -284,6 +284,8 @@ public String toString() {
   private Schema lazyColumnProjection() {
     Collection<String> selectedColumns = context.selectedColumns();
     if (selectedColumns != null) {
+      // select columns only works for data table scans, so it is OK to select on table.schema().
+
```

Comment:
I don't understand why select only works for data table scans. I updated this method to use the `schema` that is passed into the constructor and all Iceberg tests pass. By updating this to use the schema passed in instead of the table schema, we can support either project or select, depending on what the caller chooses.

Then we can remove the overrides that throw `UnsupportedOperationException`.

Reply:
The problem is on `rowFilter`.
IIUC, `rowFilter` is based on table data for every scans, so:
- if it is data table scan, we should add the field ids of `rowFilter` to `requiredFieldIds`.
- if it is not data table scan, we should not, because the schema not contains fields of `rowFilter`.